components/script/dom/bindings/js.rs

```diff
@@ -391,6 +392,158 @@ impl<T: DomObject> HeapSizeOf for MutNullableJS<T> {
     }
 }
 
+/// A holder that allows to lazily initialize the value only once 
+/// `JS<T>`, using OnceCell
+/// Essentially a `OnceCell<Option<JS<T>>>`.
+///
+/// This should only be used as a field in other DOM objects; see warning
+/// on `JS<T>`.
+#[must_root]
+pub struct OnceCellJS<T: DomObject> {
+    ptr: OnceCell<Option<JS<T>>>,
+}
+
+impl<T: DomObject> OnceCellJS<T> {
+    /// Create a new `OnceCellJS` with value
+    pub fn new(initial: Option<&T>) -> OnceCellJS<T> {
+        debug_assert!(thread_state::get().is_script());
+        OnceCellJS {
+            ptr: OnceCell::new_with_value(initial.map(JS::from_ref)),
+        }
+    }
+
+    /// Retrieve a copy of the current inner value. If it is `None`, it is
+    /// initialized with the result of `cb` first.
+    pub fn or_init<F>(&self, cb: F) -> Root<T>
+        where F: FnOnce() -> Root<T>
+    {
+        debug_assert!(thread_state::get().is_script());
+        match self.get() {
+            Some(inner) => inner,
+            None => {
+                let inner = cb();
+                self.set(Some(&inner));
+                inner
+            },
+        }
+    }
+
+    /// Retrieve a copy of the inner optional `JS<T>` as `LayoutJS<T>`.
+    /// For use by layout, which can't use safe types like Temporary.
+    #[allow(unrooted_must_root)]
+    pub unsafe fn get_inner_as_layout(&self) -> Option<LayoutJS<T>> {
+        debug_assert!(thread_state::get().is_layout());
+        match self.ptr.as_ref() {
+            Some(ptr_ref) => {
+                ptr::read(ptr_ref).map(|js| js.to_layout())
+            },
+            None => None
+        }
+        
+    }
+
+    /// Get a rooted value out of this object
+    #[allow(unrooted_must_root)]
+    pub fn get(&self) -> Option<Root<T>> {
```

Comment:
I don't think this is the right return type. The whole idea behind the issue is that whenever I call the `get` method, I shouldn't need to worry about whether the cell is empty or not, and that it will always get the appropriate value for me.